frankenstein/distributor.go

```diff
@@ -0,0 +1,102 @@
+package frankenstein
+
+import (
+	"sync"
+
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/prometheus/common/model"
+)
+
+type distributor struct {
+	consul     ConsulClient
+	ring       *Ring
+	cfg        DistributorConfig
+	clientsMtx sync.RWMutex
+	clients    map[string]storage.SampleAppender
+
+	quit chan struct{}
+	done sync.WaitGroup
+}
+
+type DistributorConfig struct {
+	ConsulHost   string
+	ConsulPrefix string
+}
+
+type collector struct {
+	hostname string
+	tokens   []uint64
+}
+
+func NewDistributor(cfg DistributorConfig) (*distributor, error) {
+	consul, err := NewConsulClient(cfg.ConsulHost)
+	if err != nil {
+		return nil, err
+	}
+	d := &distributor{
+		consul: consul,
+		ring:   NewRing(),
+		cfg:    cfg,
+		quit:   make(chan struct{}),
+	}
+	d.done.Add(1)
```

Comment:
With a max waitgroup count of 1 I'd probably just use a channel (`defer close(d.done)` in `loop()` and `<-d.done` in `Stop()`).